sql/threadpool_common.cc

```diff
@@ -275,6 +277,9 @@ static THD* threadpool_add_connection(CONNECT *connect, void *scheduler_data)
 
   thd->skip_wait_timeout= true;
   set_thd_idle(thd);
+#ifdef WITH_WSREP
+  wsrep_open(thd);
```

Comment:
Does not look good to me. It is a repetition, You have the same snippet added into do_handle_one_connection(),
and to the threadpool, where in fact you could just have it in a single place
#ifdef WITH_WSREP
    wsrep_open(thd);
#endif /* WITH_WSREP */
could be at the end of thd_prepare_connection(), rather than in both   threadpool_add_connection() and do_handle_one_connection(). 

#ifdef WITH_WSREP
    wsrep_close(thd);
#endif /* WITH_WSREP */

could be at the end_connection, rather in 2 places where it is now.

wsrep_wait_rollback_complete_and_acquire_ownership(thd); should probably also be somewhere else, e.g in
store_globals

Do not add the same code to do_handle_one_connection and threadpool

Reply:
Moved `wsrep_open()`/`wsrep_close()` into `thd_prepare_connection()`/`end_connection()` as suggested.

However, call to `wsrep_wait_rollback_complete_and_acquire_ownership()` is more tricky. The purpose of the call is to make sure that the background rollback thread is done with rollback for the `thd` object before the client thread is allowed to access `thd` data (`THD::store_globals()` in particular). The problem here is that `THD::store_globals()` is called also from background rollbacker, and there does not seem to be any mechanism in place to know whether the `THD::store_globals()` is called from client thread or from rollbacker thread. Therefore I'd like to keep the call in `threadpool_common.cc` until better solution is found. Moved the call to into `thread_attach()` to make sure that it is called in all the cases.